PurrSoft/PurrSoft.Application/Commands/VolunteerCommands/VolunteerCommandHandler.cs

```diff
@@ -0,0 +1,133 @@
+                CreatedAt = DateTime.UtcNow,
+                UpdatedAt = DateTime.UtcNow
+            };
+
+            _volunteerRepository.Add(volunteer);
+            await _volunteerRepository.SaveChangesAsync(cancellationToken);
+
+            return CommandResponse.Ok();
+        }
+        catch (Exception ex)
+        {
+            _logRepository.LogException(LogLevel.Error, ex);
+            throw;
+        }
+    }
+
+    public async Task<CommandResponse> Handle(UpdateVolunteerCommand request, CancellationToken cancellationToken)
+    {
+        CurrentUser currentUser = await _currentUserService.GetCurrentUser();
+
+        if (currentUser == null)
+        {
+            throw new UnauthorizedAccessException();
+        }
+
+        ApplicationUser user = await _userRepository
+            .Query(u => u.Id == currentUser.UserId)
+            .Include(u => u.UserRoles)
+            .ThenInclude(ur => ur.Role)
+            .FirstOrDefaultAsync(cancellationToken);
+
+        List<string> userRoles = user.UserRoles
+            .Select(ur => ur.Role.Name).ToList();
+
+        if (userRoles != null &&
+            userRoles.Contains("Volunteer") &&
+            currentUser.UserId != request.VolunteerDto.UserId)
+        {
+            throw new UnauthorizedAccessException();
+        }
+
+        try
+        {
+            Volunteer volunteer = _volunteerRepository
+                .Query(v => v.UserId == request.VolunteerDto.UserId).FirstOrDefault();
+            if (volunteer == null)
+            {
+                return CommandResponse.Failed(new[] { "There is no Volunteer with that Id!" });
+            }
+
+            volunteer.UserId = request.VolunteerDto.UserId;
+            volunteer.StartDate = DateTime.SpecifyKind(DateTime.Parse(request.VolunteerDto.StartDate), 
+                                                        DateTimeKind.Utc);
+            volunteer.EndDate = request.VolunteerDto.EndDate != null ? 
+                                DateTime.SpecifyKind(DateTime.Parse(request.VolunteerDto.EndDate), 
+                                                    DateTimeKind.Utc) : null;
+            volunteer.Status = Enum.Parse<VolunteerStatus>(request.VolunteerDto.Status);
+            volunteer.Tier = Enum.Parse<TierLevel>(request.VolunteerDto.Tier);
+            volunteer.AssignedArea = request.VolunteerDto.AssignedArea;
+            volunteer.ProfilePictureUrl = request.VolunteerDto.ProfilePictureUrl;
+            volunteer.Bio = request.VolunteerDto.Bio;
+            volunteer.UpdatedAt = DateTime.UtcNow;
+
+            await _volunteerRepository.SaveChangesAsync(cancellationToken);
+
+            return CommandResponse.Ok();
+        }
+        catch (Exception ex)
+        {
+            _logRepository.LogException(LogLevel.Error, ex);
+            throw;
+        }
+    }
+
+    public async Task<CommandResponse> Handle(DeleteVolunteerCommand request, CancellationToken cancellationToken)
+    {
+        try
```

Comment:
Why is the `UnauthorizedAccessException` return necessary in the UpdateVolunteerCommand Handler but not in the Delete Handler?

Reply:
Because the Update API call can be called by a Manager or only the Volunteer that is logged in, he should not be authorized to access this method for other Volunteers so i throw this exception to catch it in the Controller and return Forbid() back to the client. For the Delete API call i only authorized the Manager to use it (because a Volunteer should not be able to delete his account directly) and he has access to all Volunteers, i do not have to do that verification with current user logged in.